Feat/eip 4361 1 (#3)
spruceid/EIPs: EIPS/eip-4361-1.md

---
eip: TBD
title: SIWE CapGrok Extension
description: 
author: Oliver Terbu (@awoie), Jacob Ward, Charles Lehner, Sam Gbafa, Wayne Chang (@wyc), Gregory Rocco (@obstropolos)
discussions-to: TBD
status: Draft
type: Standards Track
category: ERC
created: 2021-07-20
requires: 4361
---

## Abstract

> Sign-In with Ethereum (SIWE) describes how Ethereum accounts authenticate with off-chain services (relying parties) by signing a standard message format parameterized by scope, session details, and security mechanisms (e.g., a nonce). This specification describes a mechanism on top of SIWE to give informed consent to delegate capabilities with a certain extensible scope mechanism to an authorized delegee. The way how a delegee authenticates against the target resource is out of scope of this specification and depends on the implementation of the target resource. This specification is fully-backwards compatible with EIP-4361.

Reviewer: "Backwards" compatible suggests this is a newer version of SIWE.
```suggestion
Sign-In with Ethereum (SIWE) describes how Ethereum accounts authenticate with off-chain services (relying parties) by signing a standard message format parameterized by scope, session details, and security mechanisms (e.g., a nonce). This specification describes a mechanism on top of SIWE to give informed consent to delegate capabilities with a certain extensible scope mechanism to an authorized delegee. The way how a delegee authenticates against the target resource is out of scope of this specification and depends on the implementation of the target resource. This specification is fully compatible with EIP-4361.
```

## Motivation

(non-normative)

> While SIWE focuses on authenticating the Ethereum account against the service (relying party or SIWE client) initiating the SIWE flow, there is no canonical way to interact with a third-party service (resource service) on behalf of the authenticated Ethereum account. For example, a service might want to interact with a (decentralized) protocol that is performining data storage or key management on behalf of the Ethereum account. This specification introduces a mechanism, that allows the service (or more generally a delegee) to combine authentication and authorization of such while preserving security and optimizing UX.

Reviewer: ```suggestion
While SIWE focuses on authenticating the Ethereum account against the service (relying party or SIWE client) initiating the SIWE flow, there is no canonical way to interact with a third-party service (resource service) on behalf of the authenticated Ethereum account. For example, a relying party might want to interact with another service on behalf of the Ethereum account, for example a service that provides data storage for the Ethereum account. This specification introduces a mechanism, that allows the service (or more generally a delegee) to combine authentication and authorization of such while preserving security and optimizing UX.
```

Note, this approach is a similar mechanism to combining OpenID Connect (SIWE auth) and OAuth2 (SIWE CapGrok). Resource service implementer's should not consider CapGroks as bearer tokens but instead require to authenticate the delegee in addition. The process of authenticating the delegee against the resource service is out of scope of this specification and can be done in various different ways.

> SIWE CapGroks unlock integration of (decentralized) protocols and/or APIs for developers by reducing user friction, onchain state and increasing security by introducing informed consent and deterministic capability objects.

Reviewer: ```suggestion
SIWE CapGroks unlock integration of protocols and/or APIs for developers by reducing user friction, onchain state and increasing security by introducing informed consent and deterministic capability objects.
```

This specification has three different audiences:

> - Web3 application developers that want to integrate CapGroks to interact with Web3 protocols and APIs.
> - Web3 protocol or API developers that want to learn how to define their own CapGroks.
> - Wallet implementers that want to improve the UI for capgroks.

Reviewer: In my opinion "decentralised" or "Web3" protocols/APIs is an unnecessary narrowing of the scope.
```suggestion
- Web3 application developers that want to integrate CapGroks to authenticate with any protocols and APIs that support object capabilities.
- Protocol or API developers that want to learn how to define their own CapGroks.
- Wallet implementers that want to improve the UI for CapGroks.
```

- CapGrok Translation Algorithm
- CapGrok Verification
- CapGrok Namespace Registry

### CapGrok SIWE Extension

A CapGrok is a EIP-4361 message following a specific format that allows an Ethereum account to delegate a set of CapGrok Capabilities to a delegee through informed consent. Each CapGrok Capability MUST be represented by an entry in the `Resources` array of the SIWE message that MUST deterministically translate the CapGrok Capability in human-readable form to the `statement` field in the SIWE message using the CapGrok Translation Algorithm.

The following SIWE message fields are used to further define (or limit) the scope of all CapGrok Capabilities:
- The `URI` field MUST specify the delegee as a URI, e.g., `https://example.com`, `did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK`. It is expected that the RS authenticates the delegee before invoking an action for the CapGrok Capability.
- The `Issued At` field MUST be used to specify the issuance date of the CapGrok Capabilities.
- If present, the `Expiration Time` field MUST be used as the expiration time of the CapGrok Capabilities where the RS no longer accepts invoking the associated actions.

> - If present, the `Not Before` field MUST be used as the time that has to expire before the RS starts accepting invoking of the associated actions of the CapGrok Capabilities.

Reviewer: These fields are optional in the SIWE message usually. To be clear, do we want to enforce these fields to be _required_ for a CapGrok SIWE?

Author: nope, that is why there is this `If present` language. i can make it more clear.

Reviewer: No sorry that is clear, I just skim-read it and miss the `If present`

The following is a non-normative example of a SIWE message with the SIWE CapGrok Extension:
```
example.com wants you to sign in with your Ethereum account:
0x0000000000000000000000000000000000000000

I further authorize https://example.com to perform the following actions on my behalf: (1) example: read for any. (2) example: append, delete for my.resource.1. (2) append for my.resource.2, my.resource.3.

URI: https://example.com
Version: 1
Chain ID: 1
Nonce: n-0S6_WzA2Mj
Issued At: 2022-06-21T12:00:00.000Z
Resources:
- urn:capability:example:eyJkZWZhdWx0QWN0aW9ucyI6WyJyZWFkIl0sInRhcmdldGVkQWN0aW9ucyI6eyJteS5yZXNvdXJjZS4xIjpbImFwcGVuZCIsImRlbGV0ZSJdLCJteS5yZXNvdXJjZS4yIjpbImFwcGVuZCJdLCJteS5yZXNvdXJjZS4zIjpbImFwcGVuZCJdfX0
```

### CapGrok Capability

A CapGrok Capibility is identified by their CapGrok URI that resolves to a CapGrok Details Object which defines the associated actions and optional target resources. The scope of each CapGrok Capability is attenuated by common fields in the SWIE message as described in the previous chapter, e.g., `URI`, `Issued At`, `Expiration Time`, `Not Before`.

#### CapGrok URI Scheme

> A CapGrok URI starts with `urn:capability:` followed by the namespace discriminator, followed by `:` and the base64url-encoded payload of the CapGrok Details Object.

Reviewer: I think we should define a format for namespaces. In my impl I used the following regex:
`^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$`,
i.e. any string of alpha-numeric and `-` characters, that do not start or end with a `-` and do not contain more than one consecutive `-`.

The following is a non-normative example of a CapGrok Capability that uses the `example` namespace:
```
urn:capability:example:eyJkZWZhdWx0QWN0aW9ucyI6WyJyZWFkIl0sInRhcmdldGVkQWN0aW9ucyI6eyJteS5yZXNvdXJjZS4xIjpbImFwcGVuZCIsImRlbGV0ZSJdLCJteS5yZXNvdXJjZS4yIjpbImFwcGVuZCJdLCJteS5yZXNvdXJjZS4zIjpbImFwcGVuZCJdfX0
```

It is expected that RS implementers define their own namespace, e.g., `urn:capability:service:`. It is RECOMMENDED that RS implementers use the registry table in this specification to register their namespace and associated i8n translation file to avoid name collisions and/or name squatting.

#### CapGrok Details Object Schema

> The CapGrok Details Object denotes which actions on which resources the delegee is authorized to invoke on behalf of the delegee for the validity period defined in the SIWE message. A CapGrok Details Object MUST follow the following JSON Schema:

Reviewer: ```suggestion
The CapGrok Details Object denotes which actions on which resources the delegee is authorized to invoke on behalf of the delegee for the validity period defined in the SIWE message. It can also contain additional information that the RP may require to verify a capability invocation. A CapGrok Details Object MUST follow the following JSON Schema:
```

```

> {
>    "$schema":"http://json-schema.org/draft-04/schema#",
>    "type":"object",
>    "properties":{
>       "defaultActions":{
>          "type":"array",
>          "items":{
>             "type":"string"
>          },
>          "minItems":1
>       },
>       "targetedActions":{
>          "type":"object",
>          "patternProperties":{
>             "^.*$":{
>                "type":"array",
>                "items":{
>                   "type":"string"
>                },
>                "minItems":1
>             }
>          },
>          "minProperties":1
>       }
>    },
>    "minProperties":1,
>    "additionalProperties": false
> }

Reviewer: Add `extraFields` to the JSON schema:
```suggestion
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "defaultActions": {
      "type": "array",
      "items": {
        "type": "string",
        "minLength": 1
      },  
      "minItems": 1
    },  
    "targetedActions": {
      "type": "object",
      "patternProperties": {
        "^.*$": {
          "type": "array",
          "items": {
            "type": "string",
            "minLength": 1
          },  
          "minItems": 1
        }   
      },  
      "minProperties": 1
    },  
    "extraFields": {
      "type": "object",
      "minProperties": 1
    }   
  },  
  "minProperties": 1,
  "additionalProperties": false,
  "dependentSchemas": {
    "extraFields": {
      "minProperties": 2
    }   
  }
}
```

Reviewer: Also improves validation for existing definitions:
- `defaultActions` actions must have a minimum length of 1.
- `targetedActions` targets must have a minimum length of 1.
- `targetedActions` actions must have a minimum length of 1.

Reviewer: Should we also restrict the character set for actions and targets?

```

A CapGrok Details Object defines the following properties:
- `defaultActions`: (CONDITIONAL) If present, `defaultActions` MUST be a JSON array of string values with at least one entry where each value describes an action the delegee MAY invoke in the RS on behalf of the Ethereum account without tying the scope to a particular target.
- `targetedActions`: (CONDITIONAL) If present, `targetedActions` MUST be a JSON object with variable properties where each property is a JSON array of string values each describing an action the delegee MAY invoke in the RS on behalf of the Ethereum account on the target resource denoted by the property name.

> The following is a non-normative example of a CapGrok Capability Object with both, `defaultActions` and `targetedActions`:

Reviewer: ```suggestion
The following is a non-normative example of a CapGrok Capability Object with `defaultActions`, `targetedActions` and `extraFields`:
```

```

> {
>    "defaultActions":[
>       "read"
>    ],
>    "targetedActions":{
>       "my.resource.1":[
>          "append",
>          "delete"
>       ],
>       "my.resource.2":[
>          "append"
>       ],
>       "my.resource.3":[
>          "append"
>       ]
>    }
> }

Reviewer: ```suggestion
{
   "defaultActions":[
      "read"
   ],
   "targetedActions":{
      "my.resource.1":[
         "append",
         "delete"
      ],
      "my.resource.2":[
         "append"
      ],
      "my.resource.3":[
         "append"
      ]
   },
   "extraFields":{
       "parentCapability": "bafybeigk7ly3pog6uupxku3b6bubirr434ib6tfaymvox6gotaaaaaaaaa"
   }
}
```

```

> In the example above, the delegee is authorized to perform the action `read` independent of any resource, `append`, `delete` on resource `my.resource.1`, `append` on resource `my.resource.2` and `append` on `my.resource.3`. Note, the delegee can invoke each action invididually and independent from each other in the RS.

Reviewer: ```suggestion
In the example above, the delegee is authorized to perform the action `read` independent of any resource, `append`, `delete` on resource `my.resource.1`, `append` on resource `my.resource.2` and `append` on `my.resource.3`. Note, the delegee can invoke each action invididually and independent from each other in the RS. Additionally the CapGrok Capability Object contains some additional information that the RS will need during verification. The responsibility for defining the structure and semantics of this data lies with the RS.
```

It is expected that RS implementers define which resources they want to expose through CapGrok Details Objects and which actions they want to allow to invoke on them.

### CapGrok Translation Algorithm

After applying the CapGrok Translation Algorithm on a given SIWE message that MAY include a pre-defined `statement`, the `capgrok-transformed-statement` in a CapGrok SIWE message MUST conform to the following ABNF:
```
capgrok-transformed-statement = statement capgrok-preamble 1*(" " capgrok-statement-entry ".")
   ; see EIP-4361 for definition of input-statement

> capgrok-preamble = "I further authorize " uri " to perform the following actions on my behalf:"

Author: we might need to replace constants with rule to support i8n

The following SIWE message fields are used to further define (or limit) the scope of all CapGrok Capabilities:
- The `URI` field MUST specify the delegee as a URI, e.g., `https://example.com`, `did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK`. It is expected that the RS authenticates the delegee before invoking an action for the CapGrok Capability.
- The `Issued At` field MUST be used to specify the issuance date of the CapGrok Capabilities.
- If present, the `Expiration Time` field MUST be used as the expiration time of the CapGrok Capabilities where the RS no longer accepts invoking the associated actions.
- If present, the `Not Before` field MUST be used as the time that has to expire before the RS starts accepting invoking of the associated actions of the CapGrok Capabilities.

The following is a non-normative example of a SIWE message with the SIWE CapGrok Extension:
```
example.com wants you to sign in with your Ethereum account:
0x0000000000000000000000000000000000000000

> I further authorize https://example.com to perform the following actions on my behalf: (1) example: read for any. (2) example: append, delete for my.resource.1. (2) append for my.resource.2, my.resource.3.

Reviewer: A couple of typos in `(3)`
```suggestion
I further authorize https://example.com to perform the following actions on my behalf: (1) example: read for any. (2) example: append, delete for my.resource.1. (3) example: append for my.resource.2, my.resource.3.
```

URI: https://example.com
Version: 1
Chain ID: 1
Nonce: n-0S6_WzA2Mj
Issued At: 2022-06-21T12:00:00.000Z
Resources:
- urn:capability:example:eyJkZWZhdWx0QWN0aW9ucyI6WyJyZWFkIl0sInRhcmdldGVkQWN0aW9ucyI6eyJteS5yZXNvdXJjZS4xIjpbImFwcGVuZCIsImRlbGV0ZSJdLCJteS5yZXNvdXJjZS4yIjpbImFwcGVuZCJdLCJteS5yZXNvdXJjZS4zIjpbImFwcGVuZCJdfX0
```

### CapGrok Capability

> A CapGrok Capibility is identified by their CapGrok URI that resolves to a CapGrok Details Object which defines the associated actions and optional target resources. The scope of each CapGrok Capability is attenuated by common fields in the SWIE message as described in the previous chapter, e.g., `URI`, `Issued At`, `Expiration Time`, `Not Before`.

Reviewer: ```suggestion
A CapGrok Capability is identified by their CapGrok URI that resolves to a CapGrok Details Object which defines the associated actions and optional target resources. The scope of each CapGrok Capability is attenuated by common fields in the SWIE message as described in the previous chapter, e.g., `URI`, `Issued At`, `Expiration Time`, `Not Before`.
```
